### DISCUSSION TITLE
How to get started?

### DISCUSSION THREAD

**Discussion by amijsul** [noctule-ide/noctule-swift-plugin#20] [General]
Hi all, very excited about this project!
I downloaded it through IntelliJ and am on the latest version (0.12.0.83-232)

I just don't see how to get started using it. When I opened a swift project whenever I am trying to find definitions or see usages it gives a popup error saying : "Cannot find declaration to go to" or this
<img width="579" alt="Screenshot 2023-07-27 at 10 15 17 AM" src="https://github.com/jansorg/swift-plugin/assets/135886033/e55e9153-8cb1-493d-b484-c2a0c644af27">

I have Xcode installed and have been developing just fine for a while so I don't think I'm missing any swift capabilities on my computer.

So I'm asking since there isn't a startup guide (I can help to write one if you tell me the steps!)
but how do I get the plugin to start working in IntelliJ?

IntelliJ version - 
IntelliJ IDEA 2023.2 (Community Edition)
Build #IC-232.8660.185, built on July 26, 2023
Runtime version: 17.0.7+7-b1000.6 aarch64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
macOS 13.5
GC: G1 Young Generation, G1 Old Generation
Memory: 2048M
Cores: 10
Metal Rendering is ON
Registry:
    editor.focus.mode.color.light=5a5d63
    ide.editor.tab.selection.animation=true
    external.system.auto.import.disabled=true
    debugger.watches.in.variables=false
    debugger.new.tool.window.layout=true
    ide.animate.toolwindows=true
    ide.text.editor.with.preview.show.floating.toolbar=false
    ide.instant.shutdown=false
    ide.new.editor.tabs.vertical.borders=true
    ide.experimental.ui=true
    ide.balloon.shadow.size=0
    editor.focus.mode.color.dark=5a5d63

Non-Bundled Plugins:
    dev.j-a.swift (0.12.0.83-232)
    izhangzhihao.rainbow.brackets (2023.3.2)
    com.chrisrm.idea.MaterialThemeUI (8.9.1)

Kotlin: 232-1.9.0-IJ8660.185

**Comment by jansorg** [comment]
You're right, it's currently difficult to learn how to get started. I'll prepare a guide soon. Currently, everything is still in an early state.

First of all, make sure that you configured the toolkit location (screenshot taken on Linux). It's basically the same path you used in AppCode/CLion Swift:
![image](https://github.com/jansorg/swift-plugin/assets/11473063/e4aa799e-84cb-47f2-bb11-65cfd51d2484)

Compile the project if you haven't done that already. sourcekit-lsp, which is used, currently needs this. I hope to add automatic building to simplify this step.

If navigation is still not working after reopening the project, then it may be caused by a bug. Currently, parsing errors in the source or target file break navigation. I'll fix the known parser issues soon. There's an parsing error in your screenshot, so it may be the cause. 
Could you try to find usages for `ServiceError` or go to declaration for `Error`?

**Reply by amijsul** [reply]
That worked and now it's navigating. 
Yes it appears. 
And agree that when there is an error in the file then the navigation also is broken.
So far it looks like the plugin isn't able to handle when "=" are added in certain places:

`public enum Test {
    case abc(message: String)
}`

works fine but 

`public enum Test {
    case abc(message: String = "")
}`

throws this error -> ')', ',', '.' or < expected, got '='

I got the same in a few other places where defaults are given.

Here's another weird one - it seems to be an issue with Any
this throws an error -> ',', ':', <class member>, '=' or '}' expected, got '='
`var params = [String: Any]()
`
but this doesn't
`var params = [String: String]()
`

**Reply by jansorg** [reply]
Thanks! I'll fix these errors. Swift has a complex grammar :)

**Reply by amijsul** [reply]
Awesome, and again, just the prospect of this plugin (and how it correctly renamed super fast) gets me really excited!
I don't know much about the deeper swift concepts but I'd love to contribute if you're willing to help a newbie.
I understand the language and I like it (as long as it's not being written in Xcode 😢 ) so I'd love to help contribute to fill the gap that AppCode left. Let me know!

**Comment by jansorg** [comment]
I've fixed the parsing bugs for 0.14, which should be out soon (it's not in 0.13 released today).
Please let me know how how it's working when 0.14 is available.

**Comment by jansorg** [comment]
Please note that the plugin is closed-source (as mentioned at https://github.com/jansorg/swift-plugin), so contributing to the sources is not possible, unfortunately. Spending this much time on the plugin wouldn't be possible for me, otherwise.

**Comment by jansorg** [comment]
@amijsul The parsing errors should be fixed with the current release.
Because there's close to zero feedback by users so far, how do you like the current state of the plugin, if I may ask?
Would you be sad if it became unavailable? What's the most important missing feature?
Thanks!

**Reply by amijsul** [reply]
I haven't been able to really use it yet since I have large swift projects. As I step through files I see different errors keeping it from working, I would love to be able to actually use the plugin on these larger projects, the bit I've seen so far looks good and works better than Xcode already

**Comment by ronniebarker** [comment]
Getting started still seems quite impossible. As someone who has now been using the defunct AppCode for over a year (with lots of switching to xcode and back), I still can't see how I start using this plugin in CLion to work.

With my existing xcode project, what folder/file should I open in CLion for it to start giving even the basic code highlighting and completion? Nothing seems to happen. There is a plugin message stating where the toolchain is, but nothing else seems to happen...

I purchased the plugin because I need a long-term solution to no AppCode (and xCode really isn't the solution) but it seems impossible to start and you still don't seem to have provided a decent egtting started page. You've closed the source, but that just means that you aren't getting support from what seems to be quite a large commmunity that wants this to work!

**Reply by ronniebarker** [reply]
I get the feeling that your use case is not building ios apps - but from what I can see in the AppCode discussions, that is where the most of the need for this plugin comes from. "swift build" doesn't work in an xcode project - so how does this community get started here - at the moment xcode is ticing more boxes (which is sad)

**Reply by jansorg** [reply]
Progress of this plugin is slower than I hoped, but I'm working on it.
Currently, it assumes a swift build setup, but I'll look into a way to make sourcekit-lsp work for an iOS setup.

**Reply by humblehacker** [reply]
@jansorg In case you're not already aware of this, [xcode-build-server](https://github.com/SolaWing/xcode-build-server) enables sourcekit-lsp support for Xcode projects. I've used it successfully in neovim. Perhaps it could help you add Xcode support?

Also, Wojciech Kulik has done a ton of great work building an [Xcode plugin for neovim](https://github.com/wojciech-kulik/xcodebuild.nvim). Maybe that repo could give you some ideas or maybe you could reach out to the author.

I'd really like to see your project succeed. Although I was able to get a pretty good setup going in neovim using the above, I still find myself going back to AppCode. So your plugin with CLion would be a big help.

**Reply by jansorg** [reply]
@humblehacker Thank you for your support!
I've posted a status update at https://github.com/jansorg/swift-plugin/discussions/38 and would appreciate your feedback.
